File: R_Code/Analyst_Files/staff_files_to_ray.py
import os
import glob
import re
import datetime
import win32com.client as win32
import pypandoc
import subprocess

base_path = r"C:\Users\ssanders\Documents\GitHub\my_obsidian_vault\Work\Analyst_Team"
output_dir = r"C:\Users\ssanders\Documents\GitHub\my_obsidian_vault\Work\pdfs_for_ray"

# Ensure output directory exists
os.makedirs(output_dir, exist_ok=True)

# Get all subdirectories (equivalent to list.dirs(base_path)[-1L])
# We want immediate subdirectories
input_dirs = [os.path.join(base_path, d) for d in os.listdir(base_path) 
              if os.path.isdir(os.path.join(base_path, d))]

files_to_delete = glob.glob(os.path.join(output_dir, "*.pdf"))

if files_to_delete:
    print("Deleting old .pdf files...")
    for file_path in files_to_delete:
        try:
            os.remove(file_path)
        except OSError as e:
            print(f"Error deleting {file_path}: {e}")

for input_dir in input_dirs:
    if not os.path.exists(input_dir):
        print(f"Error: Directory {input_dir} does not exist.")
        continue

    print("\nSearching for .md files in:", input_dir)
    md_files = glob.glob(os.path.join(input_dir, "*.md"))

    if not md_files:
        print("No .md files found in the specified directory.")
        continue

    print(f"Found {len(md_files)} .md files:")
    for file in md_files:
        print(f"- {os.path.basename(file)}")

    print("\n")
    
    # Find newest file based on modification time
    newest_md = max(md_files, key=os.path.getmtime)
    print("The newest file is:", newest_md)
    print("With basename of:", os.path.basename(newest_md), "\n")

    print("Converting file to PDF File.\n")
    
    pdf_filename = re.sub(r'\.md$', '.pdf', os.path.basename(newest_md))
    print("New File: ", pdf_filename)
    pdf_filepath = os.path.join(output_dir, pdf_filename)
    
    with open(newest_md, 'r', encoding='utf-8') as f:
        md_text = f.read()
        
    # Replace the checkmark emoji with [OK] to avoid LaTeX unicode errors
    md_text = md_text.replace("✅", "[OK]")
    
    try:
        # Convert using pypandoc.convert_text instead of convert_file
        pypandoc.convert_text(
            source=md_text,
            to='pdf',
            format='md',
            outputfile=pdf_filepath
        )
        print("Successfully converted to PDF.\n")
    except Exception as e:
        error_msg = str(e)
        match = re.search(r"File `(.*?)'\.sty' not found", error_msg)
        
        if match:
            missing_pkg = match.group(1)
            print(f"Missing LaTeX package detected: {missing_pkg}")
            print(f"Attempting to install {missing_pkg} via tlmgr...")
            try:
                subprocess.run(["tlmgr", "install", missing_pkg], check=True)
                print(f"Successfully installed {missing_pkg}. Retrying conversion...")
                
                # Retry conversion after installing package
                pypandoc.convert_text(
                    source=md_text,
                    to='pdf',
                    format='md',
                    outputfile=pdf_filepath
                )
                print("Successfully converted to PDF.\n")
            except subprocess.CalledProcessError as sub_e:
                print(f"Failed to install package {missing_pkg}: {sub_e}\n")
        else:
            print(f"Error during conversion: {e}\n")

output_files = glob.glob(os.path.join(output_dir, "*.pdf"))

# The most Recent Monday from today
today = datetime.date.today()
# weekday() returns 0 for Monday, so we subtract weekday() days
last_monday = today - datetime.timedelta(days=today.weekday())

filtered_paths = []
date_pattern = re.compile(r'\d{4}-\d{2}-\d{2}')

print("Extracted dates:")
print(f"{'Date':<15} | {'Kept':<6} | {'File'}")
print("-" * 50)

for file_path in output_files:
    filename = os.path.basename(file_path)
    match = date_pattern.search(filename)
    
    if match:
        date_str = match.group(0)
        try:
            file_date = datetime.datetime.strptime(date_str, "%Y-%m-%d").date()
            kept = file_date >= last_monday
            
            print(f"{file_date} | {str(kept):<6} | {filename}")
            
            if kept:
                filtered_paths.append(file_path)
                
        except ValueError:
             print(f"Invalid date format found in file: {filename}")

print("\nFiltered file paths (date >= last_monday):")
for p in filtered_paths:
    print(p)

if filtered_paths:
    try:
        print("\nPreparing email...")
        outlook = win32.Dispatch('outlook.application')
        mail = outlook.CreateItem(0)
        
        mail.To = 'Raymond.Gross2@stonybrookmedicine.edu'
        mail.Subject = 'Weekly Staff Updates'
        mail.Body = """Hi Ray,
    
Please see attached files which are the latest files I have from my 1:1 with staff.
"""
        
        # Add attachments
        for file in filtered_paths:
            # Outlook requires absolute paths
            abs_path = os.path.abspath(file)
            mail.Attachments.Add(abs_path)
            
        # Use mail.Display() to open the draft, or mail.Send() to send it immediately
        mail.Send()
        print("Email sent successfully.")
        
    except Exception as e:
        print(f"Error sending email: {e}")
else:
    print("\nNo files met the date criteria. Email not sent.")

File: R_Code/Analyst_Files/staff_files_to_ray.R
# Libraries ----
library(tidyverse)
library(RDCOMClient)
library(rmarkdown)
library(tinytex)

# Directory ----
base_path <- "C:/Users/ssanders/Documents/GitHub/my_obsidian_vault/Work/Analyst_Team/"
input_dirs <- list.dirs(base_path)[-1L]
output_dir <- "C:/Users/ssanders/Documents/GitHub/my_obsidian_vault/Work/pdfs_for_ray/"
input_dir_tbl <- tibble(
  input_dir = input_dirs,
  output_dir = output_dir
)

# Delete files in output directory ----
files_to_delete <- list.files(
  path = output_dir,
  #pattern = "\\.docx$",
  pattern = "\\.pdf$"
  full.names = TRUE
)

if (length(files_to_delete) > 0) {
  cat("Deleting old .docx files...\n")
  file.remove(files_to_delete)
}

# Process the files
input_dir_tbl |>
  group_split(input_dir) |>
  # get the newest file for each input directory ----
  imap(
    .f = function(obj, .id) {
      input_dir = obj$input_dir

      # Check if the directory exists
      if (!dir.exists(input_dir)) {
        stop(paste0("Error: Directory", input_dir, "does not exist."))
      }

      # List all .md files in the directory ----
      cat("\n")
      cat("Searching for .md files in:", input_dir, "\n")
      md_files <- list.files(
        path = input_dir,
        pattern = "\\.md$",
        full.names = TRUE
      )

      # Check if any .md files were found
      if (length(md_files) == 0) {
        stop("No .md files found in the specified directory.")
      }

      cat("Found", length(md_files), ".md files:\n")
      for (file in md_files) {
        cat("-", basename(file), "\n")
      }

      # Read in the newest .md file
      cat("\n")
      newest_md <- md_files[which.max(file.info(md_files)$mtime)]
      cat("The newest file is:", newest_md, "\n")
      cat("With basename of:", basename(newest_md), "\n\n")

      # Render file to .pdf
      cat("Converting file to Word File.\n\n")

      docx_file <- sub("\\.md$", ".pdf", basename(newest_md))
      cat("New File: ", docx_file, "\n")
      pdf_output <- render(
        input = newest_md,
        output_format = "pdf_document",#"word_document",
        output_file = paste0(obj$output_dir, docx_file)
      )

      cat("Successfully converted to Word Document.")
      cat("\n")
    }
  )

# Email out files using RDCOMClient ----
output_files <- list.files(
  path = output_dir,
  pattern = "\\.docx$",
  full.names = TRUE
)

# The most Recent Monday from today
today <- Sys.Date()
last_monday <- today - wday(today, week_start = 1) + 1

# Filter out the output_files if the date is less thatn last_monday
date_pattern <- "\\d{4}-\\d{2}-\\d{2}"
extracted_dates <- as.Date(
  regmatches(output_files, regexpr(date_pattern, output_files)),
)

# Build logical index
keep_idx <- extracted_dates >= last_monday

# Apply the filter
filtered_paths <- output_files[keep_idx]

# Print results
cat("Extracted dates:\n")
print(data.frame(
  date = extracted_dates,
  kept = keep_idx,
  file = basename(output_files)
))

cat("\nFiltered file paths (date > last_monday):\n")
print(filtered_paths)

# Create email
Outlook <- COMCreate("Outlook.Application")
Email <- Outlook$CreateItem(0)

# Set the recipient, subject, and body
Email[["to"]] = "Raymond.Gross2@stonybrookmedicine.edu"
Email[["cc"]] = ""
Email[["bcc"]] = ""
Email[["subject"]] = "Weekly Staff Updates"
Email[["body"]] = "
Hi Ray,

Please see attached files which are the latest files I have from my 1:1 with staff.
"
for (file in filtered_paths) {
  Email[["attachments"]]$Add(normalizePath(file))
}

# Send email
Email$Send()
